omegaconf/grammar_parser.py

```diff
@@ -19,11 +19,13 @@
 # Build regex pattern to efficiently identify typical interpolations.
 # See test `test_match_simple_interpolation_pattern` for examples.
 _id = "[a-zA-Z_]\\w*"  # foo, foo_bar, abc123
-_dot_path = f"(\\.)*({_id}(\\.{_id})*)?"  # foo, foo.bar3, foo_.b4r.b0z
-_inter_node = f"\\${{\\s*{_dot_path}\\s*}}"  # node interpolation
+_config_key = f"({_id}|\\$)+"  # foo, $bar, $foo$bar$
+_dot_path_res = f"({_id}(\\.{_id})*)?"  # foo, foo.bar3, foo_.b4r.b0z
+_dot_path_node = f"(\\.)*({_config_key}(\\.{_config_key})*)?"  # foo, .foo.$bar
+_inter_node = f"\\${{\\s*{_dot_path_node}\\s*}}"  # node interpolation
 _arg = "[a-zA-Z_0-9/\\-\\+.$%*@]+"  # string representing a resolver argument
 _args = f"{_arg}(\\s*,\\s*{_arg})*"  # list of resolver arguments
-_inter_res = f"\\${{\\s*{_dot_path}\\s*:\\s*{_args}?\\s*}}"  # resolver interpolation
+_inter_res = f"\\${{\\s*{_dot_path_res}\\s*:\\s*{_args}?\\s*}}"  # resolver interp.
```

Comment:
This is becoming quiet complex.
Why do we need both _dot_path_res and _dot_path_node?
res stands for resolver? if so, rename it - this is very confusing. (specially how the definition for it is intertwined with the dotpath definition.

Reply:
Yes, "res" stands for "resolver". I was trying to make sure it would fit in a single line because it gets worse when black starts making these multi-line. I tried to use better names and a more logical definition order in c3403df.

We need both `_node_path` and `_resolver_path` (new names in that commit) because we want to allow $ in node names, but not in resolver names.